New Post Tracker

EricBAndrews opened the pull request:
<!-- 
Thank you for making a pull request! 
Since we are very busy with getting Mlem into a releaseable state, we had to introduce this short questionnaire to help us review PRs.
Before you submit your PR, please take a few minutes to fill out all the needed information.

Please note that if you do not fill out the checklist, your PR will be automatically rejected unless you are an approved contributor. 
We apologize, as we would love to dedicate the time it deserves to every PR, but at present, we are under considerable time pressure.
-->

# Checklist
- [x] I have read [CONTRIBUTING.md](./CONTRIBUTING.md)
- [x] I have described what this PR contains
- [x] This PR addresses one or more open issues that were assigned to me:
      - Progress towards #73 
- [x] If this PR alters the UI, I have attached pictures/videos

# Pull Request Information

This PR rewrites the `PostTracker` to inherit from the generic `StandardTracker` and moves the saved feed out of `UserView` and into `FeedsView`. It also refactors `FeedsView` to have a cleaner navigation pattern.

## Note To Reviewers

This PR touches a massive amount of the codebase, but doesn't change much behavior. Almost all of the significant new or modified code is contained within `StandardPostTracker`, `PostFeedView`, `FeedsView`, and the generic trackers. Many of the other diffs are simple name or type changes.

## New Behavior

### Trackers

Generic trackers now have support for cursor-based pagination and have been lightly refactored to support this and improve readability.

### PostModel

PostModel is now a class, and all of its logic has been moved into it. This also allows `PostTracker` to be removed from many views where it was passed in exclusively to handle post interactions.

### Filtering

Previously, the `PostTracker`'s filter was a single function created at initialization that returned filter reasons for bookkeeping. Now, the `PostTracker` maintains a list of filters to apply that can be dynamically modified; newly fetched posts are evaluated against every tracked filter, the number of filtered items is recorded, and the filtered list is returned from the `fetch` methods.

### FeedType

`FeedType` has been fully decoupled from `APIListingType`. `FeedType` now exclusively refers to an internal feed type, and `APIListingType` matches the API specification.

### Navigation

Several unused navigation routes have been removed.

## Todo

The saved feed currently does not contain saved comments. I will open a followup PR soon to address this.

Tap tab to navigate has been removed per Slack. I have opened issue #875 to track this.

boscojwho commented:
Looks like community list view (`CommunityFeedView`?) isn't properly insetted under the tab bar.

EricBAndrews commented:
> Nice! I'll do a proper review of this tomorrow. A couple of things I noticed during some quick testing:
> 
>     * The feed refreshes when switching tabs in `CommunityView`.
> 
>     * If you set Saved as your default feed in Settings and then re-launch the app, the Saved feed appears but the ellipsis menu in the top-right isn't shown. I also had this happen once when navigating to Saved from the community list, but haven't been able to repro it.



> Nice! I'll do a proper review of this tomorrow. A couple of things I noticed during some quick testing:
> 
>     * The feed refreshes when switching tabs in `CommunityView`.
> 
>     * If you set Saved as your default feed in Settings and then re-launch the app, the Saved feed appears but the ellipsis menu in the top-right isn't shown. I also had this happen once when navigating to Saved from the community list, but haven't been able to repro it.

Looks like those both came from the same bug in initialization. Should work now.

Sjmarf commented:
Now when I open the app, I get an error haptic and it crashes immediately. I'm still using Saved as the default feed. Not sure if that affects it

It only does it about  half of the time on WiFi, but happens every single time when I use (pretty bad) mobile data 🤔 Possibly related to internet speed?

Sjmarf commented on Mlem/API/Models/ListingType.swift:
```diff
@@ -11,42 +11,5 @@ enum APIListingType: String, Codable {
-    
-    // TODO: 0.17 deprecation remove this initialiser
-    
-    init(from decoder: Decoder) throws {
```
We can't remove this decoder just yet. 

Inside of `MyUserView` we have to decode a `ListingType` in inteher form on pre-0.18.0 versions. This decoder handles that. We can't just exclude the property entirely, unfortunately - due to a bug in 0.18.0 we need to send the `ListingType` when we save the user's settings - even if they aren't modifying that setting. This bug is fixed in 0.19.0.

If we remove this a decoding error is thrown and the feed won't load on test-mlem.jo.wtf (which runs 0.17.3)

Sjmarf commented on Mlem/Models/Content/Post Model.swift:
```diff
@@ -5,26 +5,35 @@
+    var community: CommunityModel
+    @Published var votes: VotesModel
+    var commentCount: Int
+    var unreadCommentCount: Int
```
Do we want the unread comment count to disappear once you've clicked on a post? `unreadCommentCount` needs to be `@Published` if so

Sjmarf commented on Mlem/Views/Tabs/Feeds/Components/NoPostsView.swift:
In the "Saved" feed, sort-mode-related prompts can still be shown. We shouldn't do this. Repro:

- Switch to "Top of today" and enable "hide read"
- Go to saved posts
- Observe no posts message

EricBAndrews replied:
Saved feed needs its own view--what's in there right now is a stub for navigation and such. As soon as the bugs are squashed on this PR I'm going to get to work on that, so until then as long as saved is stable and the navigation works that's all that's required

Sjmarf commented on Mlem/Views/Tabs/Feeds/Components/NoPostsView.swift:
Also, this view flashes briefly before the loading animation appears when switching sort modes in the feed.

EricBAndrews replied:
Fixed

Sjmarf commented on Mlem/Views/Tabs/Feeds/Components/PostFeedView+Logic.swift:
```diff
@@ -0,0 +1,22 @@
+import SwiftUI
+
+extension PostFeedView {
+    func setDefaultSortMode() async {
```
The default-sort-mode-picking logic appears to be broken - Mlem incorrectly uses the default sort mode rather than the fallback on <0.19 instances.

EricBAndrews replied:
That *should* be fixed now

Sjmarf replied:
Nice, looks like it's working now 👍

Sjmarf replied:
Nice, looks like it's working now 👍

Sjmarf reviewed (changes requested):
Code looks good :D I found a couple of bugs

Sjmarf commented:
We had a fade-in transition on the posts in the feed before this PR. Would it be possible to bring that back? We still have it in `UserView`, but not in the feeds

EricBAndrews commented:
> We had a fade-in transition on the posts in the feed before this PR. Would it be possible to bring that back? We still have it in `UserView`, but not in the feeds

Nice catch! Added it back in

boscojwho commented:
<img width="548" alt="Screenshot 2024-01-23 at 6 53 31 PM" src="https://github.com/mlemgroup/mlem/assets/2549615/f72a732a-cf3a-4a80-af2d-f081fb47ee50">

There appears to be a missing call to `.handleLemmyLinkResolution(...)`.

Sjmarf commented on Mlem/Views/Tabs/Feeds/Components/PostFeedView.swift:
```diff
@@ -0,0 +1,167 @@
+    var content: some View {
+        LazyVStack(spacing: 0) {
+            if postTracker.items.isEmpty || versionSafePostSort == nil {
+                noPostsView()
```
Thoughts on adding a little more padding here?

```suggestion
                noPostsView()
                    .padding(.top, 50)
```

Sjmarf commented on Mlem/Views/Tabs/Feeds/Feed Types/AggregateFeedView.swift:
```diff
@@ -0,0 +1,169 @@
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                }
+                .padding(.vertical, 5)
+                    
```
Can we pad this a little more to make it consistent with the community view? 
```suggestion
                        .padding(.bottom, 3)
```

Sjmarf reviewed (changes requested):
I get a small flicker when loading feeds in certain situations, now. I think this was introduced since last review. See videos

https://github.com/mlemgroup/mlem/assets/78750526/b84e9bf3-44e7-466c-b920-3a549958d977

https://github.com/mlemgroup/mlem/assets/78750526/aa6afafc-c084-453c-ba80-89a88dad2529


I still get that single-tap haptic whenever I switch accounts, or when first opening the app.

Sjmarf commented:
The moderator badge doesn't appear to be shown in the post feed anymore

Sjmarf commented on Mlem/Views/Tabs/Feeds/Feed Types/SavedFeedView.swift:
Switching to "Saved" via the drop-down in the feed header causes the `noPostsView` to appear briefly whilst the posts are transitioning in. This only happens on the Saved feed.

Sjmarf commented on Mlem/Views/Tabs/Feeds/Feed Types/CommunityFeedView.swift:
Posts aren't refreshed when switching to "Hot" sort mode via the "Switch to hot" `noPostsView` button. 

Repro:

- Set "top of hour" as default
- Visit a community with no posts in last hour
- Tap "switch to hot"
- See "no posts" message. This shouldn't happen - instead, the posts should be shown
- Switch to "About" tab
- Switch back to "Posts" tab
- Posts load correctly

EricBAndrews commented:
> The moderator badge doesn't appear to be shown in the post feed anymore
> 
> And does the large "Communities" navigation title on the feeds list make sense here anymore? "Saved" isn't a community. Does "Feeds" make more sense?

"Feeds" definitely makes more sense, changed.

Also fixed the moderator flair. Turns out the CommunityModel bundled in UserModel doesn't have that info!

Sjmarf reviewed (approved):
Looks good! I was testing it today and didn't find any other issues.
